### DISCUSSION TITLE
When will netty support virtual thread of JDK21?

### DISCUSSION THREAD

**Discussion by zzy444626905** [netty/netty#13636] [Q&A]

**Comment by chrisvest** [comment]
The event loop threads should be platform threads, but you can use virtual threads for offloading message processing just fine.
The only issue is the use of thread-locals in the pooled bytebuf allocator, but the `io.netty.allocator.useCacheForAllThreads` system property is `false` by default, so those thread-locals won't actually pile up. Allocation will just be a bit slower in virtual threads.

What other support do you imagine we should have?

**Reply by franz1981** [reply]
I would like something like this:
- ability to create single threaded FJ pools to run each a (separated) Netty event loop 
- offload event loop jni blocking operations to an external thread pool (made of platform thread) parking till awaken: usually we epoll_wait when we have nothing to do, hence being slower to wake-up shouldn't be a big deal

The Netty event loop "cycle" will need to call Thread::yield periodically to favour interleaving with other virtual threads spawned by the same fork join pool which is powering the Netty loop: this will give the ability to user code to make use of virtual threads, sequentially ordered, running on the same carrier of the Netty event loop (which will be a long running virtual thread), hence without causing any thread-handoff typical of event loop + separated thread pool use cases.

**Reply by franz1981** [reply]
https://github.com/TechEmpower/FrameworkBenchmarks/pull/9800 fyi

**Comment by franz1981** [comment]
Fyi https://github.com/franz1981/Netty-VirtualThread-Scheduler
